File: 프로그래머스/3/258709. 주사위 고르기/주사위 고르기.js
function solution(dice) {
    const n = dice.length
    const half = n / 2
    let bestpick = []
    let maxWin = -1; 
    
    const getSums = (selected) => {
        let sums = [0]
        for (const dice of selected){
            const next = []
            for (const s of sums){
                for (const face of dice){
                    next.push(s + face)
                }
            }
            sums = next
        }
        
        return sums 
    }
    
    const combinations = (arr, k, start = 0, chosen = []) => {
        if (chosen.length === k){
            const remain = arr.filter(x => !chosen.includes(x))
            const sumA = getSums(chosen.map(i => dice[i]))
            const sumB = getSums(remain.map(i => dice[i]))
            sumA.sort((a,b) => a - b)
            sumB.sort((a,b) => a - b) 
            
            let win = 0
            // sumB는 정렬되어있음
            for (const a of sumA) {
              // sumB에서 a보다 작은 값 개수 찾기 (이진탐색)
              let left = 0, right = sumB.length;
              while (left < right) {
                let mid = Math.floor((left + right) / 2);
                if (sumB[mid] < a) left = mid + 1;
                else right = mid;
              }
              win += left; // left가 a보다 작은 값 개수
            }
            
            const total = sumA.length * sumB.length;
            const rate = win / total
            
            if (maxWin < rate){
                maxWin = rate
                bestPick = [...chosen.map(x => x + 1)]
            }
            return;
        }
        
        for (let i = start; i < n; i ++){
            chosen.push(arr[i])
            combinations(arr, k, i + 1, chosen)
            chosen.pop()
        }
    }

    combinations([...Array(n).keys()], half)
    return bestPick
}

File: 프로그래머스/3/258709. 주사위 고르기/주사위 고르기.py
from collections import Counter

def solution(dices):
    n = len(dices)
    half = n // 2
    answer = []
    total_cases = (6 ** half) ** 2
    
    def get_combinations(user_dices):
        sums = Counter()
        def dfs(depth, total):
            if depth == half:
                sums[total] += 1
                return
            for num in user_dices[depth]:
                dfs(depth + 1, total + num)
            return
        dfs(0,0)
        return sums
    
    def simulation(A_indices, B_indices):
        A_dices = [dices[i] for i in A_indices]
        B_dices = [dices[i] for i in B_indices]
        
        A_combinations = get_combinations(A_dices)
        B_combinations = get_combinations(B_dices)
        
        # 정렬후 prefix
        sorted_B = sorted(B_combinations.items())
        prefix_B = []
        total = 0
        for s, cnt in sorted_B:
            total += cnt
            prefix_B.append((s,total))
        
        # A의 각 선택에 따른 이분 탐색 시작
        A_total_wins = 0
        for s, cnt in A_combinations.items():
            left = 0
            right = len(prefix_B) - 1
            A_wins = 0
            while left <= right:
                mid = (left + right) // 2
                if prefix_B[mid][0] < s:
                    A_wins = prefix_B[mid][1]
                    left = mid + 1
                else:
                    right = mid - 1
            A_total_wins += A_wins * cnt
            
        # 케이스 점수 계산
        win_rate = A_total_wins / total_cases
        answer.append((A_indices[:], win_rate))
            
    
    def choose_A_dices(depth, start, A_indices):
        if depth == half:
            B_indices = [i for i in range(n) if i not in A_indices]
            simulation(A_indices, B_indices)
            return
        
        for i in range(start, n):
            A_indices.append(i)
            choose_A_dices(depth + 1, i + 1, A_indices)
            A_indices.pop()
        return
    
    choose_A_dices(0, 0, [])
    # 정리
    best = max(answer, key = lambda x: x[1])
    
    return sorted([num + 1 for num in best[0]])
